icu4j/main/classes/core/src/com/ibm/icu/impl/breakiter/MlBreakEngine.java

```diff
@@ -0,0 +1,434 @@
+        fDigitOrOpenPunctuationOrAlphabetSet = digitOrOpenPunctuationOrAlphabetSet;
+        fClosePunctuationSet = closePunctuationSet;
+        fModel = new HashMap<String, Integer>();
+        fNegativeSum = 0;
+        loadMLModel();
+    }
+
+    /**
+     * Divide up a range of characters handled by this break engine.
+     *
+     * @param inText        A input text.
+     * @param startPos      The start index of the input text.
+     * @param endPos        The end index of the input text.
+     * @param foundBreaks   A list to store the breakpoint.
+     * @param inString      A input string normalized from inText from startPos to endPos
+     * @param numCodePts    The number of code points of inString
+     * @param charPositions A map that transforms inString's code point index to code unit index.
+     * @return The number of breakpoint
+     */
+    public int divideUpRange(CharacterIterator inText, int startPos, int endPos,
+            CharacterIterator inString, int numCodePts, int[] charPositions,
+            DictionaryBreakEngine.DequeI foundBreaks) {
+        if (startPos >= endPos) {
+            return 0;
+        }
+        ArrayList<Integer> boundary = new ArrayList<Integer>(numCodePts);
+        int ch;
+        String ublock;
+        // The ML model groups six char to evaluate if the 4th char is a breakpoint.
+        // Like a sliding window, the elementList removes the first char and appends the new char
+        // from inString in each iteration so that its size always remains at six.
+        Element elementList[] = new Element[6];
+        initElementList(inString, elementList, numCodePts);
+
+        // Add a break for the start.
+        boundary.add(0, 0);
+        for (int i = 1; i < numCodePts; i++) {
+            evaluateBreakpoint(elementList, i, boundary);
+            if (i + 1 > numCodePts) {
+                break;
+            }
+            shiftLeftOne(elementList);
+
+            ch = (i + 3) < numCodePts ? next32(inString) : INVALID;
+            ublock = (ch != INVALID) ? getUnicodeBlock(ch) : String.valueOf((char) INVALID);
+            elementList[5].setCharAndUblock(ch, ublock);
+        }
+
+        // Add a break for the end if there is not one there already.
+        if (boundary.get(boundary.size() - 1) != numCodePts) {
+            boundary.add(numCodePts);
+        }
+
+        int correctedNumBreaks = 0;
+        int previous = -1;
+        int numBreaks = boundary.size();
+        for (int i = 0; i < numBreaks; i++) {
+            int pos = charPositions[boundary.get(i)] + startPos;
+            // In phrase breaking, there has to be a breakpoint between Cj character and close
+            // punctuation.
+            // E.g.［携帯電話］正しい選択 -> ［携帯▁電話］▁正しい▁選択 -> breakpoint between ］ and 正
+            if (pos > previous) {
+                if (pos != startPos
+                        || (pos > 0
+                        && fClosePunctuationSet.contains(inText.setIndex(pos - 1)))) {
```

Comment:
setIndex(pos - 1) returns only a 16-bit code unit, not a whole code point, so it won't work for a supplementary character. [There are 217 supplementary characters in this set](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%5B%3APc%3A%5D%5B%3APd%3A%5D%5B%3APe%3A%5D%5B%3APf%3A%5D%5B%3APo%3A%5D%5D-%5B%5Cu0000-%5CuFFFF%5D&g=&i=). Although, none of the current ones are likely to be used in Japanese text.

Please try to use
setIndex(pos) and ignore its return value and then call CharacterIteration.previous32(inText).

Reply:
Done